Retarget drop-in kubelet configuration dir feature to Alpha

sohankunkerkar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/sig node
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Retarget [this](https://github.com/kubernetes/enhancements/issues/3983) feature for Alpha, which includes:
1. Fixing the existing functionality
2. Adding an e2e test

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix overriding default KubeletConfig fields in drop-in configs if not set 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3983-drop-in-configuration
```

sohankunkerkar committed:
test/e2e_node: add e2e test for Kubeletconfig drop-in dir

Signed-off-by: Sohan Kunkerkar <sohank2602@gmail.com>
Co-authored-by: Peter Hunt <pehunt@redhat.com>

kannon92 commented:
/cc 
/ok-to-test

kannon92 commented on test/e2e_node/kubelet_config_test.go:
```diff
@@ -0,0 +1,111 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("Kubelet Config Merging with drop-in configs", func() {
```
You should follow similar pattern for naming features like we do for the other tests.

kannon92 replied:
```suggestion
var _ = SIGDescribe("Kubelet Config [NodeFeature:KubeletConfig]", func() {
```

Something like that.  We typically tag these features so its easy to search for them in the tests.

kannon92 commented on test/e2e_node/kubelet_config_test.go:
```diff
@@ -0,0 +1,111 @@
+var _ = SIGDescribe("Kubelet Config [NodeFeature:KubeletConfigDropInDir]", func() {
+	f := framework.NewDefaultFramework("kubelet-config-drop-in-dir-test")
+	ginkgo.Context("When merging kubelet configurations with drop-in configs", func() {
+		ginkgo.It("should merge kubelet configs correctly", func(ctx context.Context) {
```
What are you trying to test here?  Are all the drop-in configs relevant?  

Can we get away with just testing the `systemReserved` and drop the other fields?

sohankunkerkar replied:
>What are you trying to test here? Are all the drop-in configs relevant?

The idea is to test whether the merging of Kubelet config fields happens appropriately. It's good to test more fields with different data types to ensure the overriding logic works as expected.

kannon92 commented on test/e2e_node/kubelet_config_test.go:
```diff
@@ -0,0 +1,111 @@
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrue())
+
+			// Wait for kubelet to pick up the new configuration
```
Is there a way to know that kubelet has picked up this configuration?  Sleeps are known to be flaky in CI.

sohankunkerkar replied:
Ideally, it should pick up the moment the kubelet starts. Sleep is added to allow some breathing room.

kannon92 commented on test/e2e_node/kubelet_config_test.go:
```diff
@@ -0,0 +1,111 @@
+})
+
+// WriteKubeletDropinConfigFile writes a drop-in configuration file for the kubelet.
+func WriteKubeletDropinConfigFile(filename, dropinDir string, configData interface{}) error {
```
```suggestion
func writeKubeletDropinConfigFile(filename, dropinDir string, configData interface{}) error {
```

I don't think we need to export this function since its only used here.

sohankunkerkar replied:
Yeah, the code is still WIP. I wanted to run this test against kubernetes CI, but thanks for the suggestion.

haircommander commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

haircommander commented on test/e2e_node/kubelet_config_test.go:
```diff
@@ -0,0 +1,111 @@
+
+var _ = SIGDescribe("Kubelet Config [NodeFeature:KubeletConfigDropInDir]", func() {
+	f := framework.NewDefaultFramework("kubelet-config-drop-in-dir-test")
+	ginkgo.Context("When merging kubelet configurations with drop-in configs", func() {
```
personal nit: I'm not sure we need a context here as there aren't other Its you're grouping together

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

kannon92 commented:
/cc @kannon92

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

ffromani commented:
/triage accepted
/priority important-soon

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented on cmd/kubelet/app/server.go:
```diff
@@ -339,6 +354,23 @@ func mergeKubeletConfigurations(kubeletConfig *kubeletconfiginternal.KubeletConf
 }
 
+// convertMap recursively convert map[interface{}]interface{} to map[string]interface{}
+func convertMap(i interface{}) interface{} {
```
Let me explain my choice of using this function instead of `loadConfigFile`. I want to provide a more detailed explanation:
While running an e2e test on changes made to the kubelet config directory, I noticed a consistent test failure. The failure was due to discrepancies in the values of the following fields:
```
FileCheckFrequency
VolumeStatsAggPeriod
EvictionPressureTransitionPeriod
Authorization.Mode
EvictionHard
```
In the case of `EvictionHard`, it makes sense because we are explicitly setting the values in `loadConfigFile` if they are empty in the kubelet config. However, for the other four fields, there was no apparent reason for them to change even when we defined them in the base config. They were being overridden by default values when not explicitly set in the drop-in configs. We need to handle this in a different way to preserve the original values if they were absent in the drop-in configs. I tried investigating the code, and it appears that `DecodeKubeletConfiguration` is responsible for setting the values for kubeletconfig. However, I am uncertain if it's worth managing that logic in that function specifically for the drop-in configs, especially for those four fields.

benluddy commented on cmd/kubelet/app/server.go:
```diff
@@ -319,11 +317,28 @@ func mergeKubeletConfigurations(kubeletConfig *kubeletconfiginternal.KubeletConf
+			}
+			// Unmarshal JSON to *kubeletconfiginternal.KubeletConfiguration
+			var dropinConfig kubeletconfiginternal.KubeletConfiguration
+			if err := json.Unmarshal(data, &dropinConfig); err != nil {
```
You'll lose the distinction between absent and "present but zero-valued" here. It looks like [mergo would consider](https://github.com/kubernetes/kubernetes/blob/74fefd877f0ee9445735a756270f8eac44a96d0b/vendor/github.com/imdario/mergo/mergo.go#L36-L59) the zero values for certain types (like `metav1.Duration`) non-empty and overwrite real values.

I hacked up the fixtures for one of the unit tests in https://github.com/benluddy/kubernetes/commit/4ba879188bdc6ae55bd7110dbdc52b5e8ba1a156 to demonstrate, and it gives me:

```
--- FAIL: TestMergeKubeletConfigurations (0.00s)
    --- FAIL: TestMergeKubeletConfigurations/kubelet.conf.d_overrides_kubelet.conf (0.00s)
        server_test.go:261: 
            	Error Trace:	/var/home/benluddy/redhat/kubernetes/cmd/kubelet/app/server_test.go:261
            	Error:      	Not equal: 
            	            	expected: v1.Duration{Duration:300000000000}
            	            	actual  : v1.Duration{Duration:0}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,3 @@
            	            	 (v1.Duration) {
            	            	- Duration: (time.Duration) 300000000000
            	            	+ Duration: (time.Duration) 0
            	            	 }
            	Test:       	TestMergeKubeletConfigurations/kubelet.conf.d_overrides_kubelet.conf
            	Messages:   	Field mismatch: SyncFrequency
```

sohankunkerkar replied:
Thanks for your input. I fixed the unit test for this feature, and it's working fine now.

sohankunkerkar committed:
cmd/kubelet: fix overriding default KubeletConfig fields in drop-in configs if not set

This commit resolves an issue where certain KubeletConfig fields, specifically:
- FileCheckFrequency
- VolumeStatsAggPeriod
- EvictionPressureTransitionPeriod
- Authorization.Mode
- EvictionHard
were inadvertently overridden when not explicitly set in drop-in configs. To retain the
original values if they were absent in the drop-in configs, mergeKubeletConfigurations
uses a JSON patch merge strategy to selectively merge configurations. It prevents essential
configuration settings from being overridden, ensuring a more predictable behavior for users.

Signed-off-by: Sohan Kunkerkar <sohank2602@gmail.com>
Co-authored-by: Peter Hunt <pehunt@redhat.com>

sohankunkerkar commented:
/retest

haircommander commented on cmd/kubelet/app/server_test.go:
```diff
@@ -222,12 +232,15 @@ clusterDNS:
 			kubeletFlags := &options.KubeletFlags{}
 
-			if len(test.kubeletConfig) > 0 {
+			if !reflect.DeepEqual(test.kubeletConfig, kubeletConfig) {
```
TBH this seems kind of expensive for a check that's quite likely to be the case. Am I missing a reason to keep this?

sohankunkerkar replied:
Actually, I changed the data type for that field to `*kubeletconfiginternal.KubeletConfiguration`

sohankunkerkar replied:
Fixed

haircommander commented on test/e2e_node/services/kubelet.go:
```diff
@@ -282,6 +290,11 @@ func (e *E2EServices) startKubelet(featureGates map[string]bool) (*server, error
 	}
 
+	// Add the KubeletDropinConfigDirectory flag if set.
+	if framework.TestContext.KubeletConfigDropinDir != "" {
```
we unconditionally set this above, so I don't think we need to check this here

sohankunkerkar replied:
Agreed, fixed!

kannon92 commented:
/test pull-crio-cgroupv1-node-e2e-features

haircommander commented:
/retest

haircommander commented:
/retest

haircommander commented:
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/retest

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

kannon92 commented on test/e2e_node/kubelet_config_dir_test.go:
```diff
@@ -0,0 +1,112 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("Kubelet Config [NodeFeature:KubeletConfigDropInDir]", func() {
```
```suggestion
var _ = SIGDescribe("Kubelet Config [Slow] [Serial] [Disruptive] [NodeFeature:KubeletConfigDropInDir]", func() {
```

Since Kubelet is being restarted we should mark this as serial and disruptive.

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

sohankunkerkar commented:
/test pull-crio-cgroupv1-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

kannon92 commented:
/test

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1

kannon92 commented:
/retest

kannon92 commented:
/lgtm

There may be some renaming of these tests to catagorize them correctly.  We think this can go after the test freeze.  We run these tests locally and it doesn't look the test failures occur anymore from these tests.  

We should monitor and continue to investigate after this PR merges as the serial tests are not well covered by presubmits.

/assign @mrunalp @dims

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-containerd-alpha-features

haircommander commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mrunalp commented:
Can you squash the two test commits?

mrunalp commented on cmd/kubelet/app/server.go:
```diff
@@ -312,30 +313,34 @@ is checked every 20 seconds (also configurable with a flag).`,
-
+	baseKubeletConfigJSON, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return fmt.Errorf("failed to initially marshal base config: %w", err)
```
nit: drop `initially`

mrunalp commented:
/lgtm

mrunalp commented:
/assign @dims

dims commented:
/approve
/lgtm

the test and go mod changes are good

dims commented:
/milestone v1.29

applying milestone as this has exception from the release team here:
https://groups.google.com/g/kubernetes-sig-release/c/uvzNhRarf6w/m/sETDPDaIAAAJ
